README
NAME
    sqltree - hierarchical data (tree) implementation in SQL

VERSION
    0.06 (yyyy-mm-dd)

SYNOPSIS
      sqltree DRIVER TABLE [OPTIONS]

DESCRIPTION
    sqltree generates the SQL for a herarchical data (tree)
    implementation using triggers, as described here:

        http://www.depesz.com/index.php/2008/04/11/my-take-on-trees-in-sql/

    This implementation relies on a previously-defined table containing:

    *   a single primary key column

    *   a parent column that references the primary key

    *   a column to hold path data [optional]

    Several triggers are added to this previously-defined table, which
    update a new table holding in-depth tree information.

    Output from sqltree can usually be piped directly to the "sqlite3"
    or "psql" command line tools.

ARGUMENTS
    DRIVER
        Must be 'SQLite' or 'Pg'. Patches for other database systems are
        welcome.

    TABLE
        The name of the (existing) table holding the hierarchical data.
        The additional tree table will be postfixed with the value of
        the "--postfix" option.

OPTIONS
    --id, -i STR
        The primary key of the source table holding the hierarchical
        data. Defaults to "id".

    --no-comments, -C
        Do not include descriptive comments.

    --drop, -d
        Include DROP TABLE/TRIGGER statements.

    --name, -n STR
        When "--path" is used this option identifies the source column
        from which path names will be built. Defaults to "name".

    --parent_id, -I STR
        The parent column of the source table holding the hierarchical
        data. Defaults to "parent_id".

    --path, -p STR
        The destination column into which the tree path will be
        automatically calculated. This column would usually be defined
        as TEXT or VARCHAR, and should be UNIQUE.

    --postfix, -f STR
        The string postfixed to "TABLE" to name the additional table.
        Defaults to "_tree".

    --separator, -s CHAR
        The string that separates components in the "--path" column.
        Defaults to "/".

    --type, -t TYPE
        The SQL column type of the source table primary key and parent
        columns. Defaults to "INTEGER".

SEE ALSO
    SQL::Tree(3pm)

AUTHOR
    Mark Lawrence <nomad@null.net>

COPYRIGHT AND LICENSE
    Copyright (C) 2010-2021 Mark Lawrence <nomad@null.net>

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 3 of the License, or
    (at your option) any later version.



Changes
0.06 (yyyy-mm-dd)

    API / Behaviour Changes:
        - None.

    New Features:
        - None.

    Enhancements:
        - None.

    Bug Fixes:
        - Fix hard-coded table name in SQLite template

    Dependencies:
        - None.

    Packaging:
        - None.

0.05 (2021-01-28)

    Enhancements:
        - Add a --no-comments option.
        - Various code cleanups and improvements

0.05_1 (2021-01-27)

    API / Behaviour Changes:
        - SQL::Tree function is now just "generate(\%args)" and is not
          exported.
        - sqltree: revamp of all arguments and options

    Enhancements:
        - Modernizing and refactoring. Much better SQLite code.

    Dependencies:
        - Remove a bunch of DBI/DBD::SQLite and Test:: stuff (made
          development requirements)
        - File::ShareDir, OptArgs2, Template::Tiny, Test2::V0, ...

    Packaging:
        - Migrate to ExtUtils::MakeMaker away from Module::Install.

0.04 (2011-11-17)

    API Changes:
        - N/A

    New Features:
        - N/A

    Enhancements:
        - Perltidy and Podtidy all files with new .perltidyrc.

    Bug Fixes:
        - Correctly re-parent path columns on update.
        - Correctly update path columns on 'path_from' update.

0.03 (2010-08-03)

    API Changes:
        - Function 'generate' renamed to 'generate_sql_tree'.
        - Option 'type' becomes 'pktype'.

    New Features:
        - N/A

    Enhancements:
        - Documentation & code cleanups.

    Bug Fixes:
        - $table_tree.depth should always be type INT.

0.02 (2010-07-29)

    API Changes:
        - N/A

    New Features:
        - N/A

    Enhancements:
        - Compatible with Getop::Long::Descriptive 0.074.

    Bug Fixes:
        - N/A

0.01 (2010-07-28)

    Initial Release


Makefile.PL
#!/usr/bin/env perl
use strict;
use warnings;
use 5.010;
use ExtUtils::MakeMaker::CPANfile;
use File::ShareDir::Install;

install_share 'share';

WriteMakefile(
    NAME         => 'SQL::Tree',
    ABSTRACT     => 'Generate a trigger-based SQL tree implementation',
    AUTHOR       => 'Mark Lawrence <nomad@null.net>',
    VERSION_FROM => 'lib/SQL/Tree.pm',
    LICENSE      => 'gpl_3',
    EXE_FILES    => ['bin/sqltree'],
    test         => {
        RECURSIVE_TEST_FILES => 1,
    },
    META_MERGE => {
        'meta-spec' => { version => 2 },
        resources   => {
            bugtracker =>
              { web => 'https//github.com/mlawren/sqltree/issues', },
            homepage   => 'https//github.com/mlawren/sqltree',
            repository => {
                type => 'git',
                url  => 'ssh://git@github.com/mlawren/sql-tree.git',
                web  => 'https//github.com/mlawren/sqltree',
            }
        },
    },
);

package MY;
use File::ShareDir::Install qw(postamble);



lib/App/sqltree.pm
package App::sqltree;
use strict;
use warnings;
use File::ShareDir qw/dist_dir/;
use IO::Prompt::Tiny qw/prompt/;
use OptArgs qw/arg opt optargs/;
use Path::Tiny qw/path/;
use Template::Tiny;

our $VERSION = '0.0.5_2';

arg driver => (
    isa     => 'Str',
    comment => 'database driver type (SQLite | Pg)',
);

arg table => (
    isa     => 'Str',
    comment => 'the table name',
);

arg pk => (
    isa     => 'Str',
    comment => 'the primary key column name',
);

arg parent => (
    isa     => 'Str',
    comment => 'the parent column name',
);

arg type => (
    isa     => 'Str',
    comment => 'the SQL type of the primary key and parent columns',
);

opt help => (
    isa     => 'Bool',
    comment => 'print a usage message and exit',
    alias   => 'h',
    ishelp  => 1,
);

opt path => (
    isa     => 'Str',
    comment => 'the path column name',
    alias   => 'p',
);

opt path_from => (
    isa     => 'Str',
    comment => 'the source of the path column calculation',
    alias   => 'f',
);

opt order => (
    isa     => 'Str',
    comment => 'the order column name',
    alias   => 'o',
);

opt no_print => (
    isa     => 'Bool',
    comment => 'do not print the output, just return it',
    hidden  => 1,
);

sub run {
    my $opts = shift;
    my $share_dir =
      path( $Test::App::sqltree::SHARE_DIR || dist_dir('SQL-Tree') );

    $opts->{driver} ||= prompt( 'Driver:', 'SQLite' );
    my $template = $share_dir->child( $opts->{driver} . '.sql' );
    die "Unsupported driver type: $opts->{driver}\n" unless -f $template;

    $opts->{table}  ||= prompt( 'Table:',                 'items' );
    $opts->{pk}     ||= prompt( 'Primary Key Column:',    'id' );
    $opts->{parent} ||= prompt( 'Parent Column:',         'parent_id' );
    $opts->{type}   ||= prompt( 'PK/Parent Column Type:', 'integer' );

    $opts->{tree} = $opts->{table} . '_tree';

    my $tt  = Template::Tiny->new;
    my $src = $template->slurp_utf8;
    my $sql;

    $opts->{version} = $VERSION;
    $opts->{date}    = scalar localtime;

    $tt->process( \$src, $opts, \$sql );

    if ( $opts->{no_print} ) {
        return split( /[\n\s]*-+SPLIT-+[\n\s]*/m, $sql );
    }
    else {
        $sql =~ s/[\n\s]*-+SPLIT-+[\n\s]*/\n\n/mg;
        print $sql;
    }
}

1;
__END__


=head1 NAME

App::sqltree - implementation of the sqltree command

=head1 VERSION

0.0.5_2 (2013-06-12)

=head1 SYNOPSIS

    # Via OptArgs
    use OptArgs 'dispatch';
    dispatch(qw/run App::sqltree/);

    # Or directly
    use App::sqltree;

    my @sql = App::sqltree::run(
        {
            no_print => 1,
            driver   => 'SQLite',
            table    => 'items',
            pk       => 'id',
            parent   => 'parent_id',
            type     => 'integer',
        }
    );

=head1 DESCRIPTION

This is the implementation module for the L<sqltree> command.

=head1 SEE ALSO

L<OptArgs>, L<sqltree>

=head1 AUTHOR

Mark Lawrence E<lt>nomad@null.netE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2010,2013 Mark Lawrence E<lt>nomad@null.netE<gt>

This program is free software; you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by the
Free Software Foundation; either version 3 of the License, or (at your
option) any later version.



lib/SQL/Tree.pm
package SQL::Tree;
use strict;
use warnings;
use Carp ();
use File::Basename;
use File::ShareDir::Dist 'dist_share';
use File::Spec;
use Template::Tiny;

our $VERSION = '0.06';
our $INLINE  = {
    comments      => {},
    driver        => { required => 1, },
    drop          => {},
    id            => { required  => 1, },
    name          => { predicate => 1, },
    parent_id     => { required  => 1, },
    path          => { predicate => 1, },
    postfix       => { required  => 1, },
    separator     => { predicate => 1, },
    table         => { required  => 1, },
    template_file => {
        lazy    => 0,
        default => sub {
            my $self = shift;
            my $file =
              File::Spec->catfile( dist_share('SQL-Tree'),
                $self->driver . '.sqlt' );
            Carp::croak 'unsupported driver: ' . $self->driver unless -f $file;

            $file;
        },
    },
    type => { required => 1, },
    ref  => {
        default => sub {
            my $self = shift;
            my $ref  = {
                id        => $self->id,
                drop      => $self->drop,
                localtime => scalar localtime(),
                name      => $self->name,
                parent_id => $self->parent_id,
                path      => $self->path,
                program   => __PACKAGE__,
                separator => $self->has_separator
                ? $self->separator =~ s/'/''/gr
                : undef,
                table      => $self->table,
                table_tree => $self->table . $self->postfix,
                type       => $self->type,
                version    => $VERSION,
            };
        },
    },
};

sub BUILD {
    my $self = shift;
    Carp::croak 'path requires a name'
      if $self->has_path and not $self->has_name;
}

sub generate {
    my $self = shift;

    open my $fh, '<', $self->template_file or die 'open: ' . $!;
    my $sqlt = do { local $/; <$fh> };
    close $fh;

    my $tt = Template::Tiny->new( TRIM => 0, );
    $tt->process( \$sqlt, $self->ref, \my $output );
    $output;
}

### DO NOT EDIT BELOW! (generated by Class::Inline v0.0.1)
#<<<
  require Carp;our@ATTRS_UNEX=(undef);sub _dump {my$self=shift;require
  Data::Dumper;no warnings 'once';local$Data::Dumper::Indent=1;local
  $Data::Dumper::Maxdepth=1;my$x=Data::Dumper::Dumper($self);$x =~
  s/^.*?\n(.*)\n.*?\n/$1/s;Carp::carp$self ."\n".$x ."\n "}sub new {my$class=
  shift;my$self={@_ ? @_ > 1 ? @_ : %{$_[0]}: ()};map {local$Carp::CarpLevel=(
  $Carp::CarpLevel//0)+ 1;Carp::croak(
  "missing attribute SQL::Tree::$_ is required")unless exists$self->{$_}}
  'driver','id','parent_id','postfix','table','type';if (@ATTRS_UNEX){map {
  local$Carp::CarpLevel=$Carp::CarpLevel + 1;Carp::carp(
  "SQL::Tree attribute '$_' unexpected");delete$self->{$_ }}sort grep {not
  exists$INLINE->{$_ }}keys %$self}else {@ATTRS_UNEX=map {delete$self->{$_ };
  $_}grep {not exists$INLINE->{$_ }}keys %$self}bless$self,ref$class || $class
  ;$self->{'template_file'}//= $INLINE->{'template_file'}->{'default'}->($self
  );my@check=('SQL::Tree');my@parents;while (@check){no strict 'refs';my$c=
  shift@check;push@parents,@{$c .'::ISA'};push@check,@{$c .'::ISA'}}map {$_->
  BUILD()if exists &{$_.'::BUILD'}}reverse@parents;$self->BUILD()if exists &{
  'BUILD'};$self}sub __ro {my (undef,undef,undef,$sub)=caller(1);local
  $Carp::CarpLevel=$Carp::CarpLevel + 1;Carp::croak(
  "attribute $sub is read-only (value: '" .($_[1]// 'undef')."')")}sub
  comments {$_[0]->__ro($_[1])if @_ > 1;$_[0]{'comments'}// undef}sub driver {
  $_[0]->__ro($_[1])if @_ > 1;$_[0]{'driver'}}sub drop {$_[0]->__ro($_[1])if
  @_ > 1;$_[0]{'drop'}// undef}sub id {$_[0]->__ro($_[1])if @_ > 1;$_[0]{'id'}
  }sub name {$_[0]->__ro($_[1])if @_ > 1;$_[0]{'name'}// undef}sub has_name {
  exists $_[0]{'name'}};sub parent_id {$_[0]->__ro($_[1])if @_ > 1;$_[0]{
  'parent_id'}}sub path {$_[0]->__ro($_[1])if @_ > 1;$_[0]{'path'}// undef}sub
  has_path {exists $_[0]{'path'}};sub postfix {$_[0]->__ro($_[1])if @_ > 1;$_[
  0]{'postfix'}}sub ref {$_[0]->__ro($_[1])if @_ > 1;$_[0]{'ref'}//= $INLINE->
  {'ref'}->{'default'}->($_[0])}sub separator {$_[0]->__ro($_[1])if @_ > 1;$_[
  0]{'separator'}// undef}sub has_separator {exists $_[0]{'separator'}};sub
  table {$_[0]->__ro($_[1])if @_ > 1;$_[0]{'table'}}sub template_file {$_[0]->
  __ro($_[1])if @_ > 1;$_[0]{'template_file'}}sub type {$_[0]->__ro($_[1])if
  @_ > 1;$_[0]{'type'}}
#>>>
### DO NOT EDIT ABOVE! (generated by Class::Inline v0.0.1)

1;

__END__

=head1 NAME

SQL::Tree - Generate a trigger-based SQL tree implementation

=head1 VERSION

0.06 (yyyy-mm-dd)

=head1 SYNOPSIS

    use SQL::Tree;

    my $sql = SQL::Tree::generate({
        driver    => $DBI_driver_name,
        drop      => $bool,
        id        => $primary_key_name,
        name      => $source_column_name_for_path,
        parent    => $parent_column_name,
        path      => $path_column_name,
        table     => $table_name,
        type      => $primary_key_type,
    });


=head1 DESCRIPTION

B<SQL::Tree> generates a herarchical data (tree) implementation for
SQLite and PostgreSQL using triggers, as described here:

    http://www.depesz.com/index.php/2008/04/11/my-take-on-trees-in-sql/

A single subroutine is provided that returns a list of SQL statements:

=over 4

=item * generate( \%opts ) -> $str

=back

See the L<sqltree> documentation for the list of arguments and their
meanings.

=head1 SEE ALSO

L<sqltree>(1) - command line access to B<SQL::Tree>

=head1 AUTHOR

Mark Lawrence E<lt>nomad@null.netE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright 2010-2021 Mark Lawrence <nomad@null.net>

This program is free software; you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by the
Free Software Foundation; either version 3 of the License, or (at your
option) any later version.

